integration-tests/src/tests/client/process_blocks.rs

```diff
@@ -3900,6 +3902,265 @@ mod access_key_nonce_range_tests {
+        loop {
+            let height = test.env.clients[0].chain.head().unwrap().height;
+            if height >= 31 {
+                break;
+            }
+            debug!(target: "test", "======= Height {} ======", height + 1);
+            test.process_network_messages();
+
+            let block = test.env.clients[0].produce_block(height + 1).unwrap().unwrap();
+            if block.header().height() > 1 {
+                // For any block except the first, the previous block's application at each
+                // current chunk producer should have produced a chunk and distributed the chunk.
+                // Since we've processed all network messages just now, the block producer should
+                // have all chunks and able to create a block with all chunks. So we check the
+                // heights.
+                for i in 0..4 {
+                    assert_eq!(block.chunks()[i].height_created(), block.header().height());
+                }
+            }
+            // The block producer of course has the complete block so we can process that.
+            debug!(target: "test", "Processing block {} as the block producer", block.header().height());
+            test.env.process_block(0, block.clone(), Provenance::PRODUCED);
+
+            for i in 1..test.num_validators {
+                debug!(target: "test", "Processing block {} as validator #{}", block.header().height(), i);
+                let _ = test.env.clients[i].start_process_block(
+                    block.clone().into(),
+                    Provenance::NONE,
+                    Arc::new(|_| {}),
+                );
+                let mut accepted_blocks =
+                    test.env.clients[i].finish_block_in_processing(block.header().hash());
+                // Process any chunk part requests that this client sent. Note that this would also
+                // process other network messages (such as production of the next chunk) which is OK.
+                test.process_network_messages();
+                accepted_blocks.extend(test.env.clients[i].finish_blocks_in_processing());
+                assert_eq!(
+                    accepted_blocks.len(),
+                    1,
+                    "Processing of block {} failed at validator #{}",
+                    block.header().height(),
+                    i
+                );
+                assert_eq!(&accepted_blocks[0], block.header().hash());
+                assert_eq!(
+                    test.env.clients[i].chain.head().unwrap().height,
+                    block.header().height()
+                );
+            }
+        }
+
+        // With very high probability we should've encountered some cases where forwarded parts
+        // could not be applied because the chunk header is not available. Assert this did indeed
+        // happen.
+        assert!(test.num_forwards_with_missing_chunk_header > 0);
+        debug!(target: "test",
```

Comment:
Is this test finished? How are you testing that if the parts are forwarded, they are not requested again?

Reply:
Yes the test is finished. That logic is tested via the chunk_parts_that_must_be_known set in the custom fixture. We add a part to the set when forwarded, and check that a part is not in there if a client requests it.